File: packages/react/src/components/AdaptiveMapShell/panel-detents.test.ts
import { describe, expect, it } from "vitest";
import {
  anchoredCollapsedHeight,
  decidePanelDrag,
  nearestPanelDetent,
  orderPanelDetents,
  panelDetentHeight,
} from "./panel-detents";

// The prototype's numbers on its 874-pixel frame, driven and measured
// (docs/pointr-prototype-initial-sheet-2026-09-20.md §1).
describe("panel detents", () => {
  it("rests at a fifth, 54 % and 94 % of the shell", () => {
    expect(panelDetentHeight("collapsed", 874)).toBeCloseTo(174.8, 5);
    expect(panelDetentHeight("medium", 874)).toBeCloseTo(471.96, 5);
    expect(panelDetentHeight("large", 874)).toBeCloseTo(821.56, 5);
  });

  it("keeps a collapsed sheet tall enough for the handle and a header, but never over 40 %", () => {
    expect(panelDetentHeight("collapsed", 500)).toBe(112);
    expect(panelDetentHeight("collapsed", 200)).toBe(80);
  });

  it("rests the collapsed sheet on the content's peek anchor, within a quarter and three quarters", () => {
    expect(anchoredCollapsedHeight(250, 800)).toBe(266);
    expect(anchoredCollapsedHeight(40, 800)).toBe(192);
    expect(anchoredCollapsedHeight(700, 800)).toBe(576);
    expect(panelDetentHeight("collapsed", 800, { peekBottom: 250 })).toBe(266);
    expect(panelDetentHeight("collapsed", 800, { peekBottom: 0 })).toBe(160);
  });

  it("fits the content detent between collapsed and large, and reads as medium until measured", () => {
    expect(panelDetentHeight("content", 800)).toBe(432);
    expect(panelDetentHeight("content", 800, { contentHeight: 120 })).toBe(160);
    expect(panelDetentHeight("content", 800, { contentHeight: 300 })).toBe(300);
    expect(panelDetentHeight("content", 800, { contentHeight: 2000 })).toBe(
      752,
    );
  });

  it("clamps a fraction to the usable range and a height to the shell", () => {
    expect(panelDetentHeight({ fraction: 0.05 }, 800)).toBe(96);
    expect(panelDetentHeight({ fraction: 0.99 }, 800)).toBe(752);
    expect(panelDetentHeight({ height: 5000 }, 800)).toBe(752);
  });

  it("orders the offered detents by height and folds equal heights", () => {
    const ordered = orderPanelDetents(
      ["large", "collapsed", "medium", { fraction: 0.54 }],
      800,
    );
    expect(ordered).toEqual(["collapsed", "medium", "large"]);
  });

  it("snaps to the nearest detent by distance", () => {
    const detents = ["collapsed", "medium", "large"] as const;
    // From a fifth (175) of the prototype's frame: +100 stays, +160 reaches half.
    expect(nearestPanelDetent(detents, 275, 874)).toBe("collapsed");
    expect(nearestPanelDetent(detents, 335, 874)).toBe("medium");
    // From half (472): +160 stays, +200 reaches full; from full, −250 falls to half.
    expect(nearestPanelDetent(detents, 632, 874)).toBe("medium");
    expect(nearestPanelDetent(detents, 672, 874)).toBe("large");
    expect(nearestPanelDetent(detents, 572, 874)).toBe("medium");
  });
});

// The prototype's drag rule, decided once at a drag's first move (§2).
describe("decidePanelDrag", () => {
  it("leaves a sideways move to the content", () => {
    expect(
      decidePanelDrag({ dx: 12, dy: -8, atLargestDetent: false, scrollTop: 0 }),
    ).toBe("content");
  });
  it("moves the sheet in either direction below the largest detent", () => {
    expect(
      decidePanelDrag({ dx: 0, dy: -40, atLargestDetent: false, scrollTop: 0 }),
    ).toBe("sheet");
    expect(
      decidePanelDrag({
        dx: 0,
        dy: 40,
        atLargestDetent: false,
        scrollTop: 100,
      }),
    ).toBe("sheet");
  });
  it("scrolls the content at the largest detent on an upward drag", () => {
    expect(
      decidePanelDrag({ dx: 0, dy: -40, atLargestDetent: true, scrollTop: 0 }),
    ).toBe("content");
  });
  it("scrolls the content back to its top before the sheet moves down", () => {
    expect(
      decidePanelDrag({ dx: 0, dy: 40, atLargestDetent: true, scrollTop: 40 }),
    ).toBe("content");
    expect(
      decidePanelDrag({ dx: 0, dy: 40, atLargestDetent: true, scrollTop: 0 }),
    ).toBe("sheet");
  });
});

File: packages/android/src/test/java/com/kozmos/components/adaptivemapshell/PanelDetentsTest.kt
package com.kozmos.components.adaptivemapshell

import androidx.compose.ui.unit.dp
import org.junit.Assert.assertEquals
import org.junit.Test

/** The prototype's detents and drag rule (docs/pointr-prototype-initial-sheet-2026-09-20.md). */
class PanelDetentsTest {
    private val shell = 800.dp

    @Test
    fun theNamedDetentsAreAFifthFiftyFourAndNinetyFourPercent() {
        assertEquals(160f, KozmosMapPanelDetent.Collapsed.height(shell).value, 0.001f)
        assertEquals(432f, KozmosMapPanelDetent.Medium.height(shell).value, 0.001f)
        assertEquals(752f, KozmosMapPanelDetent.Large.height(shell).value, 0.001f)
    }

    @Test
    fun collapsedNeverFallsUnderTheHandleAndHeaderNorOverFortyPercent() {
        assertEquals(112f, KozmosMapPanelDetent.Collapsed.height(500.dp).value, 0.001f)
        assertEquals(80f, KozmosMapPanelDetent.Collapsed.height(200.dp).value, 0.001f)
    }

    @Test
    fun collapsedRestsOnThePeekAnchorWithinAQuarterAndThreeQuarters() {
        assertEquals(266f, KozmosMapPanelDetent.anchoredCollapsedHeight(250.dp, shell).value, 0.001f)
        assertEquals(192f, KozmosMapPanelDetent.anchoredCollapsedHeight(40.dp, shell).value, 0.001f)
        assertEquals(576f, KozmosMapPanelDetent.anchoredCollapsedHeight(700.dp, shell).value, 0.001f)
        assertEquals(266f, KozmosMapPanelDetent.Collapsed.height(shell, KozmosPanelMeasures(peekBottom = 250.dp)).value, 0.001f)
    }

    @Test
    fun contentFitsBetweenCollapsedAndLargeAndReadsAsMediumUntilMeasured() {
        assertEquals(432f, KozmosMapPanelDetent.Content.height(shell).value, 0.001f)
        assertEquals(160f, KozmosMapPanelDetent.Content.height(shell, KozmosPanelMeasures(contentHeight = 120.dp)).value, 0.001f)
        assertEquals(300f, KozmosMapPanelDetent.Content.height(shell, KozmosPanelMeasures(contentHeight = 300.dp)).value, 0.001f)
        assertEquals(752f, KozmosMapPanelDetent.Content.height(shell, KozmosPanelMeasures(contentHeight = 2000.dp)).value, 0.001f)
    }

    @Test
    fun theOfferedDetentsAreOrderedByHeightAndFolded() {
        val ordered = orderPanelDetents(
            listOf(KozmosMapPanelDetent.Large, KozmosMapPanelDetent.Collapsed, KozmosMapPanelDetent.Medium, KozmosMapPanelDetent.Fraction(0.54f)),
            shell
        )
        assertEquals(listOf(KozmosMapPanelDetent.Collapsed, KozmosMapPanelDetent.Medium, KozmosMapPanelDetent.Large), ordered)
    }

    @Test
    fun aDragSnapsToTheNearestDetentByDistance() {
        val detents = listOf(KozmosMapPanelDetent.Collapsed, KozmosMapPanelDetent.Medium, KozmosMapPanelDetent.Large)
        val frame = 874.dp
        // From a fifth (175): +100 stays, +160 reaches half; from half (472): +160 stays, +200 reaches full.
        assertEquals(KozmosMapPanelDetent.Collapsed, nearestPanelDetent(detents, 275.dp, frame))
        assertEquals(KozmosMapPanelDetent.Medium, nearestPanelDetent(detents, 335.dp, frame))
        assertEquals(KozmosMapPanelDetent.Medium, nearestPanelDetent(detents, 632.dp, frame))
        assertEquals(KozmosMapPanelDetent.Large, nearestPanelDetent(detents, 672.dp, frame))
    }

    @Test
    fun aSidewaysMoveIsTheContents() {
        assertEquals(KozmosPanelDragKind.Content, decidePanelDrag(dx = 12f, dy = -8f, atLargestDetent = false, scrollOffset = 0f))
    }

    @Test
    fun belowTheLargestDetentEitherDirectionMovesTheSheet() {
        assertEquals(KozmosPanelDragKind.Sheet, decidePanelDrag(dx = 0f, dy = -40f, atLargestDetent = false, scrollOffset = 0f))
        assertEquals(KozmosPanelDragKind.Sheet, decidePanelDrag(dx = 0f, dy = 40f, atLargestDetent = false, scrollOffset = 100f))
    }

    @Test
    fun atTheLargestDetentTheContentScrollsFirst() {
        assertEquals(KozmosPanelDragKind.Content, decidePanelDrag(dx = 0f, dy = -40f, atLargestDetent = true, scrollOffset = 0f))
        assertEquals(KozmosPanelDragKind.Content, decidePanelDrag(dx = 0f, dy = 40f, atLargestDetent = true, scrollOffset = 40f))
        assertEquals(KozmosPanelDragKind.Sheet, decidePanelDrag(dx = 0f, dy = 40f, atLargestDetent = true, scrollOffset = 0f))
    }
}
